### ISSUE TITLE
V9: The factory has not been configured with a proper connection string

### ISSUE THREAD

**Issue by Wiggee11** [skybrud/Skybrud.Umbraco.Redirects#115]
Installing Skybrud.Umbraco.Redirects stops Umbraco application from running if the database/connection hasn't been created yet.

```
An error occurred while starting the application.
InvalidOperationException: The factory has not been configured with a proper connection string.
Umbraco.Cms.Infrastructure.Persistence.UmbracoDatabaseFactory.Initialize()

InvalidOperationException: The factory has not been configured with a proper connection string.
Umbraco.Cms.Infrastructure.Persistence.UmbracoDatabaseFactory.Initialize()
System.Threading.LazyInitializer.EnsureInitializedCore<T>(ref T target, ref bool initialized, ref object syncLock, Func<T> valueFactory)
System.Threading.LazyInitializer.EnsureInitialized<T>(ref T target, ref bool initialized, ref object syncLock, Func<T> valueFactory)
Umbraco.Cms.Infrastructure.Persistence.UmbracoDatabaseFactory.EnsureInitialized()
Umbraco.Cms.Infrastructure.Persistence.UmbracoDatabaseFactory.get_SqlContext()
Umbraco.Cms.Core.Scoping.ScopeProvider.get_SqlContext()
Umbraco.Cms.Core.Scoping.Scope.get_SqlContext()
Umbraco.Cms.Infrastructure.Persistence.Repositories.Implement.RepositoryBase.get_SqlContext()
Umbraco.Cms.Infrastructure.Persistence.Repositories.Implement.KeyValueRepository.GetBaseQuery(bool isCount)
Umbraco.Cms.Infrastructure.Persistence.Repositories.Implement.KeyValueRepository.PerformGet(string id)
Umbraco.Cms.Core.Cache.NoCacheRepositoryCachePolicy<TEntity, TId>.Get(TId id, Func<TId, TEntity> performGet, Func<TId[], IEnumerable<TEntity>> performGetAll)
Umbraco.Cms.Infrastructure.Persistence.Repositories.Implement.EntityRepositoryBase<TId, TEntity>.Get(TId id)
Umbraco.Cms.Core.Services.Implement.KeyValueService.GetValue(string key)
Umbraco.Cms.Infrastructure.Migrations.Upgrade.Upgrader.Execute(IMigrationPlanExecutor migrationPlanExecutor, IScopeProvider scopeProvider, IKeyValueService keyValueService)
Skybrud.Umbraco.Redirects.Components.MigrationComponent.Initialize()
Umbraco.Cms.Core.Composing.ComponentCollection.Initialize()
Umbraco.Cms.Infrastructure.Runtime.CoreRuntime.StartAsync(CancellationToken cancellationToken)
Microsoft.Extensions.Hosting.Internal.Host.StartAsync(CancellationToken cancellationToken)
Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.Run(IHost host)
Application.Web.Program.Main(string[] args) in Program.cs
+
			=> CreateHostBuilder(args)
```

## To replicate

1. Follow the [steps](https://our.umbraco.com/documentation/Fundamentals/Setup/Install/) to create a fresh and clean Umbraco install with the CLI
2. Run the application `dotnet run` and you will get the installer screen
3. Install Skybrud.Umbraco.Redirects via NuGet `dotnet add package Skybrud.Umbraco.Redirects --version 3.0.0-alpha006`
4. Run the application `dotnet run` and build will fail

## Expected result

Skybrud.Umbraco.Redirects should wait until database/Umbraco has been created/installed before running migration plan.

**Comment by abjerner** [comment]
Thanks for reporting @Wiggee11 👍 

If you now how to fix this, you're more than welcome to submit a PR for this 😉 

Otherwise I'll try to find some time looking into this.

**Comment by Wiggee11** [comment]
I can't figure out how to determine the runtime level. It looks like the [RuntimeLevel attribute](https://our.umbraco.com/documentation/implementation/composing/#runtime-levels) has been removed from V9.

I don't know whether it needs to happen within the migration and inject the runtime state?...

**Comment by abjerner** [comment]
I mentioned this issue for a colleague for another package that we're updating to Umbraco 9. He just showed me that you via a notification handler can check the runtime level before adding the migration plan. 

I think this should fix the issue, so I will try looking into that 😉

**Comment by Wiggee11** [comment]
Oh cool! And if the runtime isn’t ready can you queue the migration up?

Look forward to seeing your implementation. :-)

**Comment by abjerner** [comment]
Hi @Wiggee11 

Besides the migration, there was also another similar issue when installing the redirects package before running Umbraco's installer. I forgot to mention the issue in the commit, but you can see what I've changed in the commit here: https://github.com/skybrud/Skybrud.Umbraco.Redirects/commit/c699a4adfe25a23f478eafbee56586bfbfa76db9

This should now be fixed in [**v3.0.0-alpha007**](https://github.com/skybrud/Skybrud.Umbraco.Redirects/releases/tag/v3.0.0-alpha007) 😉

**Comment by Wiggee11** [comment]
Sorry Anders, haven't had a chance to look at this until now.

Commit looks great and a fairly straightforward change in the end.

Will let you know if I come across any other issues when I upgrade the package, thanks!
